File: objectModel/Python/cdm/utilities/concurrent/_concurrent_semaphore.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from queue import Queue
from typing import Optional


class ConcurrentSemaphore:
    """Similar to the Semaphore but stops the execution on the coroutine level instead of the whole thread."""

    def __init__(self, permits: Optional[int] = None):
        self._permits = None
        # if permits is None, it means that the queue is unbounded.
        permits = 0 if permits is None else permits
        self.queue = Queue(maxsize=permits)

    @property
    def permits(self) -> Optional[int]:
        """Number of concurrent coroutine allowed to run at the same time."""
        return self._permits

    @permits.setter
    def permits(self, permits: Optional[int]) -> None:
        """Number of concurrent coroutine allowed to run at the same time."""
        self._permits = permits
        permits = 0 if permits is None else permits
        self.queue = Queue(maxsize=permits)

    def acquire(self):
        """Acquires a permit to keep the execution."""
        # Adds a number to occupy one position in the queue. 
        # If the queue gets full and another number is added,
        # the calling function will be blocked until a position is freeded up.
        self.queue.put(1)

    def release(self):
        """Releases a permit and allows other Promises to run."""
        # Release a position in the queue.
        self.queue.get()

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Utilities/Concurrent/ConcurrentSemaphore.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Utilities.Concurrent
{
    using System.Collections.Concurrent;

    internal class ConcurrentSemaphore
    {
        /// <summary>
        /// Number of concurrent tasks allowed to run at the same time.
        /// </summary>
        public int? Permits 
        {
            get
            {
                return permits;
            }
            set
            {
                permits = value;
                if (permits != null)
                {
                    queue = new BlockingCollection<int>((int)permits);
                }
                else
                {
                    // If permits is null, it means that the queue is unbounded.
                    queue = new BlockingCollection<int>();
                }
            }
        }

        /// <summary>
        /// Queue of tasks waiting for a permit to be available.
        /// </summary>
        private BlockingCollection<int> queue;

        /// <summary>
        /// Number of concurrent tasks allowed to run at the same time.
        /// </summary>
        private int? permits;

        public ConcurrentSemaphore(int? permits = null)
        {
            this.Permits = permits;
        }

        /// <summary>
        /// Acquires a permit to keep the execution.
        /// </summary>
        public void Acquire() 
        {
            // Adds a number to occupy one position in the queue. 
            // If the queue gets full and another number is added,
            // the calling function will be blocked until a position is freeded up.
            queue.Add(1);
        }

        /// <summary>
        /// eleases a permit and allows other tasks to run.
        /// </summary>
        public void Release()
        {
            // Release a position in the queue.
            queue.Take();
        }
    }
}
